aptos-move/block-executor/src/code_cache_global_manager.rs

```diff
@@ -0,0 +1,530 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{code_cache_global::GlobalModuleCache, explicit_sync_wrapper::ExplicitSyncWrapper};
+use aptos_types::state_store::StateView;
+use aptos_vm_environment::environment::AptosEnvironment;
+use move_vm_runtime::WithRuntimeEnvironment;
+use move_vm_types::code::WithSize;
+use parking_lot::Mutex;
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    ops::{Deref, DerefMut},
+    sync::Arc,
+};
+
+/// Raises an alert with the specified message. In case we run in testing mode, instead prints the
+/// message to standard output.
+macro_rules! alert_or_println {
+    ($($arg:tt)*) => {
+        if cfg!(any(test, feature = "testing")) {
+            println!($($arg)*)
+        } else {
+            use aptos_vm_logging::{alert, prelude::CRITICAL_ERRORS};
+            use aptos_logger::error;
+            alert!($($arg)*);
+        }
+    };
+}
+
+/// Represents the state of [GlobalModuleCache]. The following transitions are allowed:
+///   1. [State::Clean] --> [State::Ready].
+///   2. [State::Ready] --> [State::Executing].
+///   3. [State::Executing] --> [State::Done].
+///   4. [State::Done] --> [State::Ready].
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum State<T> {
+    Clean,
+    Ready(T),
+    Executing(T),
+    Done(T),
+}
+
+impl<T: Clone + Debug + Eq> State<T> {
+    /// If the state is [State::Clean] returns true, and false otherwise.
+    fn is_clean(&self) -> bool {
+        match self {
```

Comment:
nit: just use matches!(self, State::Clean)?

Reply:
I'd probably not even use the APIs that are a one-liner, we don't use it outside of here anyway - but that's up to the preference

Reply:
Removed one liners